src/delay/timer.rs

```diff
@@ -0,0 +1,193 @@
+//! Delay implementation based on general-purpose 32 bit timers.
+//!
+//! TIM2 and TIM5 are a general purpose 32-bit auto-reload up/downcounter with
+//! a 16-bit prescaler.
+
+use core::cmp::max;
+
+use cast::{u16, u32};
+use embedded_hal::blocking::delay::{DelayMs, DelayUs};
+
+use crate::{
+    bb,
+    pac::{self, RCC},
+    rcc::Clocks,
+};
+
+macro_rules! hal {
+    ($($TIM:ident: ($struct:ident, $waitfn:ident, $en_bit:expr, $apbenr:ident, $apbrstr:ident, $pclk:ident, $ppre:ident),)+) => {
+        $(
+            /// General purpose timer as delay provider
+            pub struct $struct {
+                clocks: Clocks,
+                tim: pac::$TIM,
+            }
+
+            fn $waitfn(tim: &mut pac::$TIM, prescaler: u16, auto_reload_register: u32) {
+                // Write Prescaler (PSC)
+                tim.psc.write(|w| w.psc().bits(prescaler));
+
+                // Write Auto-Reload Register (ARR)
+                // Note: Make it impossible to set the ARR value to 0, since this
+                // would cause an infinite loop.
+                tim.arr.write(|w| unsafe { w.bits(max(1, auto_reload_register)) });
```

Comment:
Note: This `max` call makes it impossible to delay for 0 ms / us, but I don't see the use case for that. It would be undefined (in terms of actual delay time) anyways.